### DISCUSSION TITLE
How to get metadata from a query without executing it?

### DISCUSSION THREAD

**Discussion by fdcastel** [sqlalchemy/sqlalchemy#10771] [Usage Questions]
Given a SQL (text) query and a connection to a database, how do you discover what columns/datatypes (e.g. `DateTime`) this query would return (without executing the entire query)?

To the best of my understanding, the _prepare_ ("compile") phase of a query execution should provide this. 

After preparation, the database should return the metadata for the query result (before any fetch) and its parameters (e.g. `ID INT, DESCRIPTION VARCHAR(100), :P1 INT, :P2 DATETIME`) or an error (e.g. `SQL error at char 42.`).

How can I use SQLAlchemy's already-existing infrastructure to accomplish this in a database-independent way?

**Comment by zzzeek** [comment]
hi - 

a *textual* query?  like a string "SELECT a, b, c FROM table"?  There's no datatype information in that string by itself.  if you are looking to have a server-side conversation with the database using prepared statements, first note that most DBAPIs don't use prepared statements at all, and for the few that do, it's generally concealed by the driver, as there is no prepared statement API in pep-249.

Overall then this question would be highly database-backend specific, and there is definitely no database-independent way to get typing metadata for a SQL string without actually invoking the statement on the server, and even then the typing information you get will vary greatly, if even present at all (such as on SQLite where you get basically nothing).

**Comment by fdcastel** [comment]
> a _textual_ query? like a string "SELECT a, b, c FROM table"? 

Exactly.

> if you are looking to have a server-side conversation with the database using prepared statements (...)

Yes!

> first note that most DBAPIs don't use prepared statements at all, and for the few that do, it's generally concealed by the driver, as there is no prepared statement API in pep-249.

Daaaaaaaaaaaaamn! 😞

> Overall then this question would be highly database-backend specific, and there is definitely no database-independent way to get typing metadata for a SQL string without actually invoking the statement on the server, and even then the typing information you get will vary greatly (...) 

I was sure I could get this info from `CursorResult` or `CusorResultMetaData`. 
Back to the planning board, then.

> (...) if even present at all (such as on SQLite where you get basically nothing).

I presume that's also why I got `None` from `.cursor.description`.

Thanks @zzzeek! Much appreciated.

**Reply by zzzeek** [reply]
SQLlite shouldnt give you "None" if you do a SELECT, but the type element of cursor.description I think is empty.   you definitely should get the cursor.description collection though.

**Reply by fdcastel** [reply]
You are correct. SQLite does indeed return `cursor.description`. My fault.

It is the _Firebird_ driver that does return `None`. I will contact the author about this.

**Reply by fdcastel** [reply]
Oh, great! Now even Firebird is returning `cursor.description` correctly (!?). I probably need some hours of sleep 😅.

```python
r = conn.exec_driver_sql(f"{query} ROWS 0")

all_columns = {c: dt for c, dt, *rest in r.cursor.description}
date_column_names = [c for c in all_columns if issubclass(all_columns[c], date)]
```

and we are done! 

Caveats:

- For SQLite, it returns only column names with "None" as the datatype. 
  - But this is acceptable given SQLite's infamous... "flexibility" with column types.

- I had to actually execute the query (not just prepare it). But this is also acceptable in my case. 

- The only problem may be that `ROWS 0`, which some databases may frown upon. 
  - Suggestions about how to use SQLAlchemy to limit this (in a dialect-independent way) are welcome 😅.

**Reply by CaselIT** [reply]
> - Suggestions about how to use SQLAlchemy to limit this (in a dialect-independent way) are welcome 😅.

not 100% sure how supported it is, but the following works
```py
import sqlalchemy as sa
q = sa.text('select foo from bar')
sa.select(q).limit(1)  # not sure if 0 may be ignored by some dialect
```

**Reply by fdcastel** [reply]
Yes. This will render a SQL with server-side limit implemented by SQLAlchemy dialect.👍

Problem is: I'm using the lower level `exec_driver_sql` to avoid problems with SQLAlchemy trying to parse `:` as parameters. 

I have some column alias which may contain `:`. E.g.:
```
  SELECT column AS "crazy:column-name" (...)
```

There is a way to disable this behaviour? If yes I could use @CaselIT solution.

_Alternatively, should this (trying to parse parameters inside SQL double quotes) be considered a bug?_

**Reply by CaselIT** [reply]
> Alternatively, should this (trying to parse parameters inside SQL double quotes) be considered a bug?

the behavior is documented and you can escape it. see https://docs.sqlalchemy.org/en/20/core/sqlelement.html#sqlalchemy.sql.expression.text
